### DISCUSSION TITLE
Difficulty implementing `SetWindowsHookEx`

### DISCUSSION THREAD

**Discussion by JordanAnthonyKing** [microsoft/CsWin32#248] [Q&A] [answered]
## Actual behavior

Apologies if I'm misunderstanding something, documentation is rather sparse. I'm attempting to implement a low level keyboard hook in a .net 5.0 console (for now) app.

Translating from a simple C++ example:

```
  s_llKeyboardHook = SetWindowsHookEx(WH_KEYBOARD_LL, LowLevelKeyboardProc, GetModuleHandle(NULL), NULL);
  
  static inline HHOOK s_llKeyboardHook = nullptr;
  
  static LRESULT CALLBACK LowLevelKeyboardProc(int nCode, WPARAM wParam, LPARAM lParam)
  {
      /* Some code */
      return CallNextHookEx(NULL, nCode, wParam, lParam);
  }
```

To C#:

class Program
```
{
  static ManualResetEvent _quitEvent = new ManualResetEvent(false);
  
  static void Main(string[] args)
  {
	  Console.CancelKeyPress += (sender, eArgs) =>
	  {
		  _quitEvent.Set();
		  eArgs.Cancel = true;
	  };
  
	  using SafeHandle handle = PInvoke.SetWindowsHookEx(SetWindowsHookEx_idHook.WH_KEYBOARD_LL, HookProc, PInvoke.GetModuleHandle((string)null), 0);
  
	  _quitEvent.WaitOne();
	  handle.Close();
  }
  
  static LRESULT HookProc(int nCode, WPARAM wParam, LPARAM lParam)
  {
	  System.Diagnostics.Debug.WriteLine("Hello World!");
	  
	  return PInvoke.CallNextHookEx(null, nCode, wParam, lParam);
  }
}
```

Attempting to use `PInvoke.GetModuleHandle((string)null),` gives the error: `CS1503	Argument 3: cannot convert from 'nint' to 'System.Runtime.InteropServices.SafeHandle'`

I've seen a few more involved examples than this, and some where just `null` is provided in place of a handler but following any of these has resulted in the HookProc not being called.

I believe this is related to https://github.com/microsoft/win32metadata/issues/147 but couldn't find a workaround in the thread.

What is the correct way to go about this?

## Expected behavior
A way to use GetModuleHandle in SetWindowsHookEx, or a working alternative.

## Repro steps

1. `NativeMethods.txt` content:
```
CreateFile
SetWindowsHookEx
GetModuleHandle
CallNextHookEx
MarshallAs
```
2. Create a new .net 5.0 console app
3. Paste the code as given into Program.cs

### Context

- CsWin32 version: `0.4.422-beta`
- Target Framework: `.net 5.0`

**Comment by AraHaan** [comment]
Cant you pass in ``null`` directly without casting to ``(string)``?

Also why use ``GetModuleHandle(null);`` over something like the way to get the handle to the executing assembly and using that from the BCL itself?

**Comment by JordanAnthonyKing** [comment]
> Cant you pass in `null` directly without casting to `(string)`?

Passing in null without casting is ambiguous.

> Also why use GetModuleHandle(null); over something like the way to get the handle to the executing assembly and using that from the BCL itself?

Most of this is pretty new to me, I've fiddled with this stuff a little via Power Toys, mostly I just want to get the hook set up and then code away as normal in C#. I attempted briefly writing a basic SafeHandle that would just return false for `isInvalid` and seeing what would happen ala this example from https://github.com/microsoft/win32metadata/issues/147 but no luck getting any output.
```
  _hCtrl = PInvoke.CreateWindowEx(Constants.WS_EX_NOPARENTNOTIFY,
      "Button",
      "&Settnigs",
      Constants.WS_TABSTOP | Constants.WS_VISIBLE | Constants.WS_CHILD | Constants.WS_GROUP,
      10, 10, 88, 24, _hDlg,
      new DestroyMenuSafeHandle((IntPtr) 3, false),
      new FreeLibrarySafeHandle(PInvoke.GetModuleHandle((string)null), false),
      null);
```

**Comment by AArnott** [comment, marked as answer]
Your problem may not be the interop at all. Maybe it's because the way the callback is invoked is via the Windows message pump, and your console app doesn't have one.

**Reply by JordanAnthonyKing** [reply]
> Your problem may not be the interop at all. Maybe it's because the way the callback is invoked is via the Windows message pump, and your console app doesn't have one.

Thank you for the answer. I will look into this.
